File: shared/types/money.ts
/**
 * Money Value Object (TypeScript)
 * 
 * Matches Go implementation in backend/pkg/money/money.go
 * - Stored as int64 cents for precision
 * - No floating-point errors
 * - Type-safe currency operations
 */
export class Money {
  constructor(
    private readonly amount: number,  // int64 cents
    private readonly currency: string
  ) {
    if (!Number.isInteger(amount)) {
      throw new Error('Amount must be an integer (cents)');
    }
    if (!currency || currency.length !== 3) {
      throw new Error('Currency must be a 3-letter code (e.g., USD, EUR)');
    }
  }

  /**
   * Get amount in cents (int64)
   */
  getAmount(): number {
    return this.amount;
  }

  /**
   * Get currency code (e.g., "USD", "EUR")
   */
  getCurrency(): string {
    return this.currency;
  }

  /**
   * Convert to float for display (e.g., 100.50)
   */
  toFloat(): number {
    return this.amount / 100;
  }

  /**
   * Format for display (e.g., "$100.50")
   */
  format(locale = 'en-US'): string {
    return new Intl.NumberFormat(locale, {
      style: 'currency',
      currency: this.currency,
    }).format(this.toFloat());
  }

  /**
   * Create Money from float (e.g., 100.50)
   */
  static fromFloat(value: number, currency: string): Money {
    if (value < 0) {
      throw new Error('Amount cannot be negative');
    }
    return new Money(Math.round(value * 100), currency);
  }

  /**
   * Create Money from cents (int64)
   */
  static fromCents(cents: number, currency: string): Money {
    return new Money(cents, currency);
  }

  /**
   * Create zero Money
   */
  static zero(currency: string): Money {
    return new Money(0, currency);
  }

  /**
   * Add two Money values
   */
  add(other: Money): Money {
    if (this.currency !== other.currency) {
      throw new Error(`Cannot add different currencies: ${this.currency} and ${other.currency}`);
    }
    return new Money(this.amount + other.amount, this.currency);
  }

  /**
   * Subtract Money
   */
  subtract(other: Money): Money {
    if (this.currency !== other.currency) {
      throw new Error(`Cannot subtract different currencies: ${this.currency} and ${other.currency}`);
    }
    return new Money(this.amount - other.amount, this.currency);
  }

  /**
   * Multiply by factor
   */
  multiply(factor: number): Money {
    if (factor < 0) {
      throw new Error('Factor cannot be negative');
    }
    return new Money(Math.round(this.amount * factor), this.currency);
  }

  /**
   * Divide by factor
   */
  divide(factor: number): Money {
    if (factor <= 0) {
      throw new Error('Factor must be positive');
    }
    return new Money(Math.round(this.amount / factor), this.currency);
  }

  /**
   * Check if zero
   */
  isZero(): boolean {
    return this.amount === 0;
  }

  /**
   * Check if negative
   */
  isNegative(): boolean {
    return this.amount < 0;
  }

  /**
   * Check if positive
   */
  isPositive(): boolean {
    return this.amount > 0;
  }

  /**
   * Compare equality
   */
  equals(other: Money): boolean {
    return this.amount === other.amount && this.currency === other.currency;
  }

  /**
   * Compare greater than
   */
  greaterThan(other: Money): boolean {
    if (this.currency !== other.currency) {
      throw new Error('Cannot compare different currencies');
    }
    return this.amount > other.amount;
  }

  /**
   * Compare less than
   */
  lessThan(other: Money): boolean {
    if (this.currency !== other.currency) {
      throw new Error('Cannot compare different currencies');
    }
    return this.amount < other.amount;
  }

  /**
   * Serialize for JSON
   */
  toJSON(): { amount: number; currency: string } {
    return {
      amount: this.amount,
      currency: this.currency,
    };
  }

  /**
   * Parse from JSON
   */
  static fromJSON(data: { amount: number; currency: string }): Money {
    return new Money(data.amount, data.currency);
  }

  /**
   * String representation
   */
  toString(): string {
    return `${this.currency} ${this.toFloat().toFixed(2)}`;
  }
}

/**
 * Currency codes
 */
export type Currency = 'USD' | 'EUR' | 'UAH' | 'GBP' | 'CAD' | 'AUD';

/**
 * Money utilities
 */
export const MoneyUtils = {
  /**
   * Sum multiple Money values (same currency)
   */
  sum(values: Money[]): Money {
    if (values.length === 0) {
      throw new Error('Cannot sum empty array');
    }
    
    const currency = values[0].currency;
    const total = values.reduce((sum, m) => {
      if (m.currency !== currency) {
        throw new Error('All Money values must have the same currency');
      }
      return sum + m.amount;
    }, 0);
    
    return new Money(total, currency);
  },

  /**
   * Parse from API response (float or cents)
   */
  parseFromAPI(value: number | { amount: number; currency: string }, currency?: Currency): Money {
    if (typeof value === 'object') {
      return Money.fromJSON(value);
    }
    
    // Assume value is float from API response
    return Money.fromFloat(value, currency || 'USD');
  },
};

File: pkg/money/money.go
// Package money provides a Money value object for handling monetary values.
// It stores amounts in the smallest currency unit (cents) to avoid floating-point precision issues.
package money

import (
	"errors"
	"fmt"
	"strings"
)

var (
	// ErrNegativeAmount is returned when attempting to create money with a negative amount.
	ErrNegativeAmount = errors.New("amount cannot be negative")
	// ErrDifferentCurrencies is returned when attempting operations on different currencies.
	ErrDifferentCurrencies = errors.New("cannot operate on different currencies")
	// ErrInvalidCurrency is returned when currency code is invalid.
	ErrInvalidCurrency = errors.New("invalid currency code: must be 3-letter ISO 4217 code")
)

// Money represents a monetary value in the smallest currency unit (cents/minor units).
// This avoids floating-point precision issues common with currency calculations.
type Money struct {
	Amount   int64  // Amount in smallest currency unit (e.g., cents for USD)
	Currency string // ISO 4217 currency code (e.g., USD, EUR, UAH)
}

// New creates a new Money instance from an amount in the smallest currency unit and a currency code.
func New(amount int64, currency string) (Money, error) {
	if amount < 0 {
		return Money{}, ErrNegativeAmount
	}
	if currency == "" {
		return Money{}, errors.New("currency is required")
	}
	currency = strings.ToUpper(currency)
	if len(currency) != 3 {
		return Money{}, ErrInvalidCurrency
	}

	return Money{
		Amount:   amount,
		Currency: currency,
	}, nil
}

// NewFromFloat creates Money from a float64 amount (e.g., 12.34 -> 1234 cents).
// This is convenient for converting user input or API responses.
func NewFromFloat(amount float64, currency string) (Money, error) {
	if amount < 0 {
		return Money{}, ErrNegativeAmount
	}
	// Multiply by 100 and round to avoid floating-point errors
	return New(int64(amount*100+0.5), currency)
}

// Add adds two Money values. Returns error if currencies don't match.
func (m Money) Add(other Money) (Money, error) {
	if m.Currency != other.Currency {
		return Money{}, ErrDifferentCurrencies
	}
	return Money{
		Amount:   m.Amount + other.Amount,
		Currency: m.Currency,
	}, nil
}

// Subtract subtracts two Money values. Returns error if currencies don't match or result would be negative.
func (m Money) Subtract(other Money) (Money, error) {
	if m.Currency != other.Currency {
		return Money{}, ErrDifferentCurrencies
	}
	if m.Amount < other.Amount {
		return Money{}, ErrNegativeAmount
	}
	return Money{
		Amount:   m.Amount - other.Amount,
		Currency: m.Currency,
	}, nil
}

// Multiply multiplies Money by a factor. Returns error if factor is negative.
func (m Money) Multiply(factor float64) (Money, error) {
	if factor < 0 {
		return Money{}, ErrNegativeAmount
	}
	result := int64(float64(m.Amount)*factor + 0.5)
	return Money{
		Amount:   result,
		Currency: m.Currency,
	}, nil
}

// Divide divides Money by a factor. Returns error if factor is zero or negative.
func (m Money) Divide(factor float64) (Money, error) {
	if factor <= 0 {
		return Money{}, errors.New("division factor must be positive")
	}
	result := int64(float64(m.Amount)/factor + 0.5)
	return Money{
		Amount:   result,
		Currency: m.Currency,
	}, nil
}

// ToFloat64 converts Money to float64 for display or API responses (e.g., 1234 cents -> 12.34).
func (m Money) ToFloat64() float64 {
	return float64(m.Amount) / 100.0
}

// String returns formatted string representation (e.g., "12.34 USD").
func (m Money) String() string {
	return fmt.Sprintf("%.2f %s", m.ToFloat64(), m.Currency)
}

// Equals checks if two Money values are equal.
func (m Money) Equals(other Money) bool {
	return m.Amount == other.Amount && m.Currency == other.Currency
}

// IsZero checks if amount is zero.
func (m Money) IsZero() bool {
	return m.Amount == 0
}

// IsPositive checks if amount is positive.
func (m Money) IsPositive() bool {
	return m.Amount > 0
}

// CompareTo compares two Money values.
// Returns: -1 if m < other, 0 if m == other, 1 if m > other
// Panics if currencies don't match.
func (m Money) CompareTo(other Money) int {
	if m.Currency != other.Currency {
		panic(fmt.Sprintf("cannot compare different currencies: %s vs %s", m.Currency, other.Currency))
	}
	if m.Amount < other.Amount {
		return -1
	} else if m.Amount > other.Amount {
		return 1
	}
	return 0
}

// GreaterThan checks if this Money is greater than other.
func (m Money) GreaterThan(other Money) bool {
	return m.CompareTo(other) > 0
}

// LessThan checks if this Money is less than other.
func (m Money) LessThan(other Money) bool {
	return m.CompareTo(other) < 0
}

// GreaterThanOrEqual checks if this Money is greater than or equal to other.
func (m Money) GreaterThanOrEqual(other Money) bool {
	return m.CompareTo(other) >= 0
}

// LessThanOrEqual checks if this Money is less than or equal to other.
func (m Money) LessThanOrEqual(other Money) bool {
	return m.CompareTo(other) <= 0
}

// GetAmount returns the amount in smallest currency unit.
func (m Money) GetAmount() int64 {
	return m.Amount
}

// GetCurrency returns the ISO 4217 currency code.
func (m Money) GetCurrency() string {
	return m.Currency
}
